goroutine leaks

stream.NewClientStream seems to leak a goroutine.

I have some unsent local changes to the end2end tests which check for goroutine leaks on each test.

I see:

```
--- FAIL: TestExceedMaxStreamsLimit (7.02s)
        end2end_test.go:365: Running test in TCP environment...
        end2end_test.go:1388: Leaked goroutine: goroutine 2738 [select]:
                google.golang.org/grpc.NewClientStream.func1(0x7f9b8946d248, 0xc82035e690, 0xc8200da2a0, 0xc820013400)
                        /home/bradfitz/src/google.golang.org/grpc/stream.go:148 +0x257
                created by google.golang.org/grpc.NewClientStream
                        /home/bradfitz/src/google.golang.org/grpc/stream.go:156 +0xa99
--- FAIL: TestCompressOK (6.06s)
        end2end_test.go:365: Running test in TCP environment...
        end2end_test.go:1388: Leaked goroutine: goroutine 2750 [select]:
                google.golang.org/grpc.NewClientStream.func1(0x7f9b8946d248, 0xc82035e870, 0xc82010c9a0, 0xc820013860)
                        /home/bradfitz/src/google.golang.org/grpc/stream.go:148 +0x257
                created by google.golang.org/grpc.NewClientStream
                        /home/bradfitz/src/google.golang.org/grpc/stream.go:156 +0xa99
FAIL
FAIL    google.golang.org/grpc/test     33.222s
```

That leak is from `stream.go`:

```go
// NewClientStream creates a new Stream for the client side. This is called
// by generated code.
func NewClientStream(ctx context.Context, desc *StreamDesc, cc *ClientConn, method string, opts ...CallOption) (ClientStream, error) {
....
....
....
        // Listen on ctx.Done() to detect cancellation when there is no pending
        // I/O operations on this stream.
        go func() {
                select {
                case <-t.Error():
                        // Incur transport error, simply exit.
                case <-s.Context().Done():
                        err := s.Context().Err()
                        cs.finish(err)
                        cs.closeTransportStream(transport.ContextErr(err))
                }
        }()
        return cs, nil
}
```

I started looking for this when running the race detector with a high `go test -race -count=NNN` value, because the race detector hit its trackable goroutine limit and crashed, which made me suspicious what was leaking.

I'll send changes for adding the goroutine leak checker.

I noticed more failures on the Travis run once I enabled this:

```
--- FAIL: TestHealthCheckOnFailure-4 (5.31s)
	end2end_test.go:356: Running test in tcp-clear environment...
	end2end_test.go:356: Running test in tcp-tls environment...
	end2end_test.go:356: Running test in unix-clear environment...
	end2end_test.go:356: Running test in unix-tls environment...
	<autogenerated>:31: Leaked goroutine: goroutine 10361 [select]:
		google.golang.org/grpc/transport.(*http2Server).controller(0xc2080e6000)
			/home/travis/gopath/src/google.golang.org/grpc/transport/http2_server.go:626 +0x84f
		created by google.golang.org/grpc/transport.newHTTP2Server
			/home/travis/gopath/src/google.golang.org/grpc/transport/http2_server.go:134 +0xb66
```

@iamqizhao, this one doesn't look as obviously easy to fix. I'm going to disable the leak checker for now and get it submitted, and then you can fix this bug.

--- a/server.go
+++ b/server.go
@@ -279,7 +279,7 @@ func (s *Server) serveNewHTTP2Transport(c net.Conn, authInfo credentials.AuthInf
 		return
 	}
 	if !s.addConn(st) {
-		c.Close()
+		st.Close()
 		return
 	}
 	s.serveStreams(st)

--- a/test/end2end_test.go
+++ b/test/end2end_test.go
@@ -1354,7 +1354,7 @@ func interestingGoroutines() (gs []string) {
 	return
 }
 
-var failOnLeaks = flag.Bool("fail_on_leaks", false, "Fail tests if goroutines leak.")
+var failOnLeaks = flag.Bool("fail_on_leaks", true, "Fail tests if goroutines leak.")
 
 // leakCheck snapshots the currently-running goroutines and returns a
 // function to be run at the end of tests to see whether any

--- a/test/end2end_test.go
+++ b/test/end2end_test.go
@@ -1354,8 +1354,6 @@ func interestingGoroutines() (gs []string) {
 	return
 }
 
-var failOnLeaks = flag.Bool("fail_on_leaks", true, "Fail tests if goroutines leak.")
-
 // leakCheck snapshots the currently-running goroutines and returns a
 // function to be run at the end of tests to see whether any
 // goroutines leaked.
@@ -1363,10 +1361,6 @@ func leakCheck(t testing.TB) func() {
 	orig := map[string]bool{}
 	for _, g := range interestingGoroutines() {
 		orig[g] = true
-	}
-	leakf := t.Logf
-	if *failOnLeaks {
-		leakf = t.Errorf
 	}
 	return func() {
 		// Loop, waiting for goroutines to shut down.
@@ -1387,7 +1381,7 @@ func leakCheck(t testing.TB) func() {
 				continue
 			}
 			for _, g := range leaked {
-				leakf("Leaked goroutine: %v", g)
+				t.Errorf("Leaked goroutine: %v", g)
 			}
 			return
 		}

